DataFormats/Provenance/src/IndexIntoFile.cc

```diff
@@ -1980,27 +2030,408 @@ namespace edm {
+        // Count entries with events
+        ++nSize;
+      }
+    }
+    info.firstIndexOfRun_.resize(maxOrderPHIDRun + 1, invalidIndex);
+    info.firstIndexOfLumi_.resize(maxOrderPHIDRunLumi + 1, invalidIndex);
+    info.startOfLastContiguousEventsInRun_.resize(maxOrderPHIDRun + 1, invalidIndex);
+    info.startOfLastContiguousEventsInLumi_.resize(maxOrderPHIDRunLumi + 1, invalidIndex);
+    info.indexesSortedByEventEntry_.reserve(nSize);
+
+    // Reserve some space. Most likely this is not big enough, but better than reserving nothing.
+    fileOrderRunOrLumiEntry_.reserve(runOrLumiEntries.size());
+    shouldProcessRunOrLumi_.reserve(runOrLumiEntries.size());
+    shouldProcessEvents_.reserve(runOrLumiEntries.size());
+  }
+
+  void IndexIntoFile::IndexIntoFileItrEntryOrder::gatherNeededInfo(EntryOrderInitializationInfo& info) {
+    auto const& runOrLumiEntries = indexIntoFile()->runOrLumiEntries();
+    int iEnd = static_cast<int>(runOrLumiEntries.size());
+
+    EntryNumber_t previousLumi = invalidEntry;
+    EntryNumber_t previousRun = invalidEntry;
+    int index = 0;
+
+    for (auto const& runOrLumiEntry : runOrLumiEntries) {
+      // If first entry for a lumi
+      if (!runOrLumiEntry.isRun() && runOrLumiEntry.orderPHIDRunLumi() != previousLumi) {
+        previousLumi = runOrLumiEntry.orderPHIDRunLumi();
+
+        // Fill map holding the first index into runOrLumiEntries for each lum
+        info.firstIndexOfLumi_[runOrLumiEntry.orderPHIDRunLumi()] = index;
+      }
+
+      // If first entry for a run
+      if (runOrLumiEntry.orderPHIDRun() != previousRun) {
+        previousRun = runOrLumiEntry.orderPHIDRun();
+
+        // Fill map holding the first index into runOrLumiEntries for each run
+        info.firstIndexOfRun_[runOrLumiEntry.orderPHIDRun()] = index;
+
+        // Look ahead to see if the run has events or not
+        bool runHasEvents = false;
+        for (int indexWithinRun = index + 1;
+             indexWithinRun < iEnd && runOrLumiEntries[indexWithinRun].orderPHIDRun() == runOrLumiEntry.orderPHIDRun();
+             ++indexWithinRun) {
+          if (runOrLumiEntries[indexWithinRun].beginEvents() != invalidEntry) {
+            runHasEvents = true;
+            break;
+          }
+        }
+        if (!runHasEvents) {
+          info.runsWithNoEvents_.push_back({runOrLumiEntry.entry(), index});
+        }
+      }
+      ++index;
+    }
+
+    std::sort(info.runsWithNoEvents_.begin(),
+              info.runsWithNoEvents_.end(),
+              [](TTreeEntryAndIndex const& left, TTreeEntryAndIndex const& right) -> bool {
+                return left.ttreeEntry_ < right.ttreeEntry_;
+              });
+
+    info.nextRunWithNoEvents_ = info.runsWithNoEvents_.cbegin();
+    info.endRunsWithNoEvents_ = info.runsWithNoEvents_.cend();
+  }
+
+  void IndexIntoFile::IndexIntoFileItrEntryOrder::fillIndexesSortedByEventEntry(EntryOrderInitializationInfo& info) {
```

Comment:
Another member function of `EntryOrderInitializationInfo` or `const`.

Reply:
Done. I moved it into the Info class. That is better. Thanks.